### PR TITLE
Handle completed FlushResult on CopyToAsync

### LINKED ISSUE #1958: PipeReader CopyToAsync to a PipeWriter doesn't notice when the destination Pipe is closed.
With .NET Core 3.1, I'm using a PipeReader.CopyToAsync to copy to a destination pipe.  When I close the pipe, the CopyToAsync doesn't notice until the buffer is full.

`await Request.BodyReader.CopyToAsync(receivePipe.Pipe.Writer);`

I would expect an exception or something.  As it is, the PipeReader.Complete seems kind of useless.

Looking in PipeReader.cs, it appears that it pays attention and throws an exception for a cancellation, but doesn't do anything about IsComplete.

            return CopyToAsyncCore(destination, async (destination, memory, cancellationToken) =>
            {
                FlushResult result = await destination.WriteAsync(memory, cancellationToken).ConfigureAwait(false);

                if (result.IsCanceled)
                {
                    ThrowHelper.ThrowOperationCanceledException_FlushCanceled();
                }
            },
            cancellationToken);

I've managed to work around this by passing along a cancellation token along with the pipe, but I feel kind of ridiculous writing:

            cancellationSource.Cancel();
            receivePipe.Reader.Complete();

This stuff is fantastic though.  Even with the occasional hiccup, it's such a huge leap forward.

### PR DESCRIPTION
**Pull request by manandre**
Fixes #1958
/cc @davidfowl @chriswelles

### CODE REVIEW

File: src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs [hunk 1]
```diff
@@ -158,13 +152,15 @@ public virtual Task CopyToAsync(Stream destination, CancellationToken cancellati
                 return Task.FromCanceled(cancellationToken);
             }
 
-            return CopyToAsyncCore(
-                destination,
-                (destination, memory, cancellationToken) => destination.WriteAsync(memory, cancellationToken),
-                cancellationToken);
+            return CopyToAsyncCore(destination, async (destination, memory, cancellationToken) =>
+            {
+                await destination.WriteAsync(memory, cancellationToken).ConfigureAwait(false);
```
Comment (davidfowl) [inline src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:157]: You can avoid the state machine if this finishes synchronously.


```suggestion
                var task = destination.WriteAsync(memory, cancellationToken).ConfigureAwait(false);
                
                static async Task Awaited(Task writeTask)
                {
                    await writeTask;
                    return new FlushResult(isCanceled: false, isCompleted: false);
                }
                
                if (task.IsCompletedSuccessfully)
                {
                    task.GetAwaiter().GetResult();
                    return new FlushResult(isCanceled: false, isCompleted: false);
                }
                return Awaited(task);
```

File: src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs [hunk 2]
```diff
@@ -182,9 +191,19 @@ private async Task CopyToAsyncCore<TStream>(TStream destination, Func<TStream, R
...
+                        FlushResult flushResult = await writeAsync(destination, memory, cancellationToken).ConfigureAwait(false);
+
+                        if (flushResult.IsCanceled)
+                        {
+                            ThrowHelper.ThrowOperationCanceledException_FlushCanceled();
+                        }
 
                         consumed = position;
+
+                        if (flushResult.IsCompleted)
```
Comment (davidfowl) [inline src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:203]: Do we have a test for this case? If you `PipeReader.CopyToAsync(PipeWriter)` and that pipe writer is complete before writing all of the data, and you copy the PipeReader to another Pipewriter and it resumes where it left off and continues copying.

Comment (manandre) [reply src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:203]: I have added a test proposal. Comments are welcome.
I have checked that without the fix this test case hangs... as expected.
However, I have a doubt around the update of the `consumed` position when the flush result is "completed". Indeed, by design, flush comes after write so data is considered as consumed (if not failed), but the [`Pipe.WriteAsync`](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L959) method can return a completed FlushResult before consuming the provided data...
Do I miss something here?

Comment (davidfowl) [reply src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:203]: That's a really good point. Lets file another issue. There's basically no way to know if the data has been consumed or not. Great find!

Comment (BrennanConroy) [reply src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:203]: You shouldn't have this check if you're doing other important stuff in this loop (which you are).

Comment (halter73) [reply src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:203]: This was written during triage when I didn't have the full context. I've now read through most of the related issues.

It seems the real problem is that when you have a PipeWriter/Reader pair with arbitrary Pause/ResumeThreshold, you cannot observe the read loop completion just using PipeWriter. This has always been the case though. Checking FlushResult.IsCompleted doesn't change that.

You used by be able to use [https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.Pipelines/ref/System.IO.Pipelines.cs#L79-L80](PipeWriter.OnReaderCompleted) and a TCS for this, but that's been obsoleted. If we want to keep this API, we'd want it to be a Task anyway.

Comment (davidfowl) [reply src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:203]: I think we leave this as it was the existing behavior. Assuming it was consumed.

**Review by halter73 — APPROVED** [review]

**Comment by davidfowl** [comment]
Thanks @manandre !

**Pull request merged**